File: Easy/0203_remove-linked-list-elements/remove-linked-list-elements.py
#!/bin/python3

from typing import List, Optional
from dataclasses import dataclass

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeElements(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        dummy = ListNode(0)
        dummy.next = head
        current = dummy

        while current.next:
            if current.next.val == val:
                current.next = current.next.next
            else:
                current = current.next
        return dummy.next

def build_linked_list(values: List[int]) -> Optional[ListNode]:
    dummy = ListNode()
    current = dummy
    for v in values:
        current.next = ListNode(v)
        current = current.next
    return dummy.next

def print_linked_list(node: Optional[ListNode]) -> None:
    while (node):
        print(node.val, end=" -> ")
        node = node.next
    print ("None")

@dataclass
class TestCase:
    l : List[int]
    v : int

def main():
    tests : List[TestCase] = [
        TestCase([1,2,6,3,4,5,6], 6),
        TestCase([], 1),
        TestCase([7,7,7,7], 7),
    ]

    for t in tests:
        l = build_linked_list(t.l)
        print(f"---\nlist: ", end='')
        print_linked_list(l)
        l = Solution().removeElements(l, t.v)
        print(f"afterRemove: ", end='')
        print_linked_list(l)

if __name__ == "__main__":
    main()

File: Easy/0203_remove-linked-list-elements/remove-linked-list-elements.go
package main

import (
	"fmt"
)

type ListNode struct {
	Val int
	Next *ListNode
}

func removeElements(head *ListNode, val int) *ListNode {
	dummy := &ListNode{Next: head}
	current := dummy

	for current.Next != nil {
		if current.Next.Val == val {
			current.Next = current.Next.Next
		} else {
			current = current.Next
		}
	}
  return dummy.Next
}

func buildLinkedList(values []int) *ListNode {
	dummy := &ListNode{}
	current := dummy
	for _, v := range values {
		current.Next = &ListNode{Val: v}
		current = current.Next
	}
	return dummy.Next
}

func printLinkedList(node *ListNode) {
	for node != nil {
		fmt.Printf("%d => ", node.Val)
		node = node.Next
	}
	fmt.Println("nil")
}

type Unit struct {
	l []int
	v int
}

func main() {
	tests := []Unit{
		{l: []int{1, 2, 6, 3, 4, 5, 6}, v: 6},
		{l: []int{}, v: 1},
		{l: []int{7, 7, 7, 7}, v: 7},
	}

	for _, t := range tests {
		fmt.Printf("---\nlist: ")
		list := buildLinkedList(t.l)
		printLinkedList(list)
		fmt.Printf("afterRemove: ")
		res := removeElements(list, t.v)
		printLinkedList(res)
		fmt.Printf("\n")
	}
}
